Sources/Commands/SwiftTestTool.swift

```diff
@@ -1046,7 +1054,13 @@ final class XUnitGenerator {
                 """
 
             if !result.success {
-                stream <<< "<failure message=\"failed\"></failure>\n"
+                let message = [testErrorRegex, fatalErrorRegex].map { $0.matchGroups(in: result.output) }
+                    .first(where: { !$0.isEmpty })?
+                    .first?
+                    .joined(separator: " ")
+                    .xmlEscaped ?? "failed"
+
+                stream <<< "<failure message=\"\(message)\"></failure>\n"
```

Comment:
this seems like an improvement over the hard coded "failed" message, but a bit fragile. @stmontgomery any advice?

Reply:
It really should not be necessary to scrape the textual logging output of XCTests and parse it using regex. We have a proper API for observing events and retrieving this data in structured ways: [XCTestObservation](https://developer.apple.com/documentation/xctest/xctestobservation).

However, I think SwiftPM needs to coordinate doing this for it to be effective but it doesn't do that today. So I can understand why this approach relying on more-and-more intricate regex parsing is seen as the easiest solution at the moment. I believe SwiftPM already parses the textual logging output to a certain extent (right?), so this is only adding a bit more and I'm not going to stand in the way of landing this. But I think we really should spend some time building the necessary infrastructure for SwiftPM to coordinate more closely with the runner process to gather its results in a structured way, to avoid the need for any of this.

@tomerd Do you know if we already have an Issue tracking this sort of enhancement on the SwiftPM side?

Reply:
@stmontgomery we have been pushing back on adding parsing of XCTest output and we would very much like to see a structured output from XCTest instead of the non-structured one we have now - it will help solve several issues. iirc the chalk age is that the invocation of XCTest goes through several layers that essentially shell out so there is little room for setting up delegates etc, but maybe there was a flag we can pass to the process that can assist? open to ideas on how to be at achieve this